lib/Driver/UnixToolChains.cpp

```diff
@@ -182,27 +182,30 @@ toolchains::GenericUnix::constructInvocation(const DynamicLinkJobAction &job,
 
   // Configure the toolchain.
   //
-  // By default use the system `clang` to perform the link.  We use `clang` for
-  // the driver here because we do not wish to select a particular C++ runtime.
-  // Furthermore, until C++ interop is enabled, we cannot have a dependency on
-  // C++ code from pure Swift code.  If linked libraries are C++ based, they
-  // should properly link C++.  In the case of static linking, the user can
-  // explicitly specify the C++ runtime to link against.  This is particularly
-  // important for platforms like android where as it is a Linux platform, the
-  // default C++ runtime is `libstdc++` which is unsupported on the target but
-  // as the builds are usually cross-compiled from Linux, libstdc++ is going to
-  // be present.  This results in linking the wrong version of libstdc++
-  // generating invalid binaries.  It is also possible to use different C++
```

Comment:
I think it would be good if we didn't lose this information about Android-specific problems.

Reply:
I felt conflicted about removing this as well -- but I'm not sure this is the best place to surface this information. (This is also why I asked @compnerd to take a look.)

This sounds to me almost more like an issue with Clang (or the way it's configured) than with Swift itself. Assuming we're passing the right `-target` to Clang (we're doing that, right?), shouldn't it select the correct `-stdlib` default for the platform? If it isn't, shouldn't we file this as a bug against Clang (and potentially put a warning in user-facing documentation until the bug is fixed) rather than putting this in a code comment?

Reply:
Sounds good to me as well. I think we should just not lose that information.

Reply:
Agree that we shouldn't lose this information.

I think the underlying issue here is that some component is exhibiting host / target confusion -- either Clang, or Swift in the way it uses Clang. Until we know which of the two it is, maybe the best thing is to do here is to open a Swift bug, e.g. "Ensure C++ interop defaults to the correct C++ standard library for the target, not the host".

That bug would reference the comment I'm deleting here. The first action item on the bug would be to work out whether this is an issue in Swift or Clang, then proceed from there.

I think that would be a better place for the information to live. For the purposes of this source file, the goal of the comment I'm deleting is really just to make it clear to maintainers that they shouldn't replace `clang` with `clang++` unconditionally, because that might have undesirable consequences. I've tried to fulfill that purpose with the comment "We don't use `clang++` unconditionally because...".

WDYT?

Let's leave this comment thread open until @compnerd has had a chance to respond.

Reply:
SG

Reply:
> I don't think that there is a good place to surface this information sadly.
> 
> The problem is even more interesting in general: a Linux distribution _can_ have multiple supported C++ runtimes.

That's understood.

> You need to not only ensure that the link is correctly referencing the correct library, but the import is also getting the correct headers as otherwise you have an ABI mismatch.

This (making sure the import is getting the correct headers) is done by passing the appropriate `-stdlib=...` flag to Clang, correct? (Specifically, the Clang used in ClangImporter; the Clang invocation that is used on some platforms to link the executable should receive the same `-stdlib` flag.)

> Passing the correct `-target` doesn't do what you want (and really cannot, it can default, but that doesn't make it correct).

I understand that the default C++ standard library may not be what the user wants -- but shouldn't Clang at least choose a default that is supported on the `-target` that was specified (even if the user may want a different C++ standard library in case the target supports more than one)?

This is what I was proposing opening a bug for. I'm basing this on the following comment in UnixToolChains.cpp:

```
  // [...] This is particularly
  // important for platforms like android where as it is a Linux platform, the
  // default C++ runtime is `libstdc++` which is unsupported on the target but
  // as the builds are usually cross-compiled from Linux, libstdc++ is going to
  // be present.  This results in linking the wrong version of libstdc++
  // generating invalid binaries.
```

IIUC, this is saying the following:

*  Android doesn't support libstdc++; executables compiled for it need to use a different standard library (presumably libc++?).
*  When using `clang++` to cross-compile for Android (target) on Linux (host) without a `-stdlib` option, Clang will choose libstdc++ as the default because this is the default on the host.

If true, this second point seems wrong. IMO, Clang's logic for choosing a default for `-stdlib` should be based on what the target is, not what the host is. Is my expectation here wrong? Are there technical reasons for why the `-stdlib` default cannot be made dependent on what the `-target` is?

> You need to pass `-stdlib=` with the correct value. I may want to build against gnustl or libc++ or even a different C++ runtime. There is no correct answer here, the user is the only one who can know for certain.
> 
> If you really want to lift this, I think that you need to expose the C++ runtime through to the driver and require that the user specifies the runtime that they are trying to use.

I've added an option `-experimental-cxx-stdlib` to the driver that allows the user to specify which standard library they want.

I haven't made this a required flag though; as Clang doesn't require `-stdlib` to be specified (and will choose a hopefully meaningful default if it isn't), I think we shouldn't either?

> I think that clang is actually correct in its behaviour here - its not a configuration issue, its a the user must tell us what they are trying to do.

Agree that the user should have this option; I was just hoping that Clang would choose a useful default (see above).

Reply:
Update: I've created https://bugs.swift.org/browse/SR-12664 to track the Android issue and capture the information there.